### DISCUSSION TITLE
Support for environment

### DISCUSSION THREAD

**Discussion by PatrickDanino** [pointfreeco/swift-perception#1] [General]
Great work overall!

One question I had while perusing is whether it would be possible to also add support for [environment(:)](https://developer.apple.com/documentation/swiftui/view/environment(_:)) and possibly other features that are also iOS 17 specific.

We currently use this to inject context into our views.

**Comment by valvoline** [comment]
Indeed. That's would be a really great addition, something like 

```
.perceptive(:)
```

BTW, not sure if this's "correct" in respect to the library, but seems to work quite well in my first tests and it seems to achieve (at least, generally speaking) to the ```.environment(:)``` needs.

```swift
import SwiftUI
import Perception

@Perceptible
class CounterModel: ObservableObject {
    var counter: Int = 0
}

struct DetailView: View {
    @EnvironmentObject private var model: CounterModel
    var body: some View {
        WithPerceptionTracking {
            VStack {
                Text("Counter: \(model.counter)")
                Button(action: {
                    model.counter += 1
                }, label: {
                    Text("Increment me")
                })
            }
        }
    }
}

struct ContentView: View {
    let model = CounterModel()
    @State private var isSheetShowing = false
    var body: some View {
        WithPerceptionTracking {
            VStack {
                Text("Counter: \(model.counter)")
                Button(action: {
                    model.counter += 1
                }, label: {
                    Text("Increment me")
                })
                Button(action: {
                    isSheetShowing.toggle()
                }, label: {
                    Text("Show Modal")
                })
            }
        }
        .sheet(isPresented: $isSheetShowing, content: {
            DetailView()
        })
        .environmentObject(model)
    }
}
```

I added the ObservableObject conformances. Not sure if this could lead to problems with the library itself.

**Reply by mbrandonw** [reply]
Hi @valvoline, what you have outlined here is actually totally fine, and can be a work around until we decide how we want to backport `environment(_:)`. Marking your class as `ObservableObject` is totally fine, and that allows you to put it in the environment, and so it effectively works the exact same as `environment(_:)`.

**Comment by mbrandonw** [comment]
Hi @PatrickDanino, we will look into back porting `environment(_:)`, but in the meantime @valvoline's suggestion above will work just fine.

**Comment by stephencelis** [comment]
@PatrickDanino @valvoline We've added a backport here: https://github.com/pointfreeco/swift-perception/pull/13

Want to try it out?

**Reply by valvoline** [reply]
Oh!, many thanks! I'll fetch it tomorrow and test. I see you used an EnvironmentValues to backport. Nice.

**Reply by valvoline** [reply]
Works quite well. Many thanks, again.
